STM32/I2C/Src/sm4291.c

```diff
@@ -0,0 +1,310 @@
+    return sm4291ReadReg(i2c, ADDR_STATUS_SYNC, status);
+}
+
+/*!
+** @brief Reset the device
+*/
+int sm4291Reset(sm4291_i2c_handle_t* i2c) {
+    return sm4291WriteReg(i2c, ADDR_CMD, CMD_RESET);
+}
+
+/*!
+** @brief Put the device into sleep mode
+**
+** @note The datasheet doesn't specify how t get the device out of sleep mode
+*/
+int sm4291Sleep(sm4291_i2c_handle_t* i2c) {
+    return sm4291WriteReg(i2c, ADDR_CMD, CMD_SLEEP);
+}
+
+/***************************************************************************************************
+** PRIVATE FUNCTIONS
+***************************************************************************************************/
+
+/*!
+** @brief Reads a 16-bit register from the device
+*/
+static int sm4291ReadReg(sm4291_i2c_handle_t* i2c, uint8_t addr, uint16_t* result) {
+    assert_param(i2c);
+
+    uint16_t dev_addr = i2c->crc ? TEMP_I2C_CRC_ADDR : TEMP_I2C_NO_CRC_ADDR;
+
+    if(i2c->crc) {
+        uint8_t crc4_data[2U] = {(uint8_t)(addr >> 4U), ((addr & 0xFU) << 4U) & 0x1U};
+        uint8_t crc4 = genCrc4(crc4_data, 2U);
+
+        uint8_t buf[2U] = {addr, (uint8_t)(0x10U | crc4)};
+        if(HAL_OK != HAL_I2C_Master_Transmit(i2c->i2c, dev_addr << 1, buf, 2, 1)) {
+            return -1;
+        };
+
+        uint8_t buf1[3U] = {0};
+        if(HAL_OK != HAL_I2C_Master_Receive(i2c->i2c, dev_addr << 1, buf1, 3, 1)) {
+            return -2;
+        };
+
+        if(genCrc8(buf1, 2) != buf1[2]) {
+
+            return -3;
+        }
+
+        *result = ((uint16_t) buf1[1] << 8U) | buf1[0];
+    }
+    else {
+        if(HAL_OK != HAL_I2C_Master_Transmit(i2c->i2c, dev_addr << 1, &addr, 1, 1)) {
+            return -1;
+        };
+
+        if(HAL_OK != HAL_I2C_Master_Receive(i2c->i2c, dev_addr << 1, (uint8_t*)result, 2, 1)) {
+            return -2;
+        };
+    }
+
+    return 0;
+}
+
+/*!
+** @brief Writes a 16-bit register to the device
+*/
+static int sm4291WriteReg(sm4291_i2c_handle_t* i2c, uint8_t addr, uint16_t reg) {
+    assert_param(i2c);
+
+    uint16_t dev_addr = i2c->crc ? TEMP_I2C_CRC_ADDR : TEMP_I2C_NO_CRC_ADDR;
+    if(i2c->crc) {
+        uint8_t crc4_data[2U] = {(uint8_t)(addr >> 4U), ((addr & 0xF) << 4U) & 0x1U};
+        uint8_t crc8_data[2U] = {(uint8_t)(reg & 0xFFU), (uint8_t)((reg >> 8U) & 0xFFU)};
+        uint8_t crc4 = genCrc4(crc4_data, 2U);
+        uint8_t crc8 = genCrc8(crc8_data, 2U);
+
+        uint8_t buf[5U] = {addr, (uint8_t)(0x10 | crc4), (uint8_t)(reg & 0xFFU), (uint8_t)((reg >> 8U) & 0xFFU), crc8};
+        return (int)HAL_I2C_Master_Transmit(i2c->i2c, dev_addr << 1, buf, 5U, 1);
+    }
+    else {
+        uint8_t buf[3U] = {addr, (uint8_t)(reg & 0xFF), (uint8_t)((reg >> 8U) & 0xFFU)};
+        return (int)HAL_I2C_Master_Transmit(i2c->i2c, dev_addr << 1, buf, 3U, 1);
+    }
+}
+
+/*!
+** @brief Calculates CRC8
+**
+** Taken from: https://stackoverflow.com/questions/51752284/how-to-calculate-crc8-in-c
```

Comment:
Maybe this comment should be added to `SHT35.c` and `SHT45.c` as they are identical :smile:

Reply:
Maybe CRC should be moved to its own file?!